pkg/render/typha.go

```diff
@@ -620,43 +620,18 @@ func (c *typhaComponent) typhaPodSecurityPolicy() *policyv1beta1.PodSecurityPoli
 	return psp
 }
 
-// affinity sets the affinity on typha, accounting for the kubernetes-provider and user-specified values.
+// affinity sets the user-specified typha affinity if specified.
 func (c *typhaComponent) affinity() (aff *v1.Affinity) {
-	// in AKS, there is a feature called 'virtual-nodes' which represent azure's container service as a node in the kubernetes cluster.
-	// virtual-nodes have many limitations, namely it's unable to run hostNetworked pods. virtual-kubelets are tainted to prevent pods from running on them,
-	// but typha tolerates all taints and will run there.
-	// as such, we add a required anti-affinity for virtual-nodes if running on azure
-	if c.installation.KubernetesProvider == operator.ProviderAKS {
-		aff = &v1.Affinity{
-			NodeAffinity: &v1.NodeAffinity{
-				RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
-					NodeSelectorTerms: []v1.NodeSelectorTerm{{
-						MatchExpressions: []v1.NodeSelectorRequirement{
-							{
-								Key:      "type",
-								Operator: corev1.NodeSelectorOpNotIn,
-								Values:   []string{"virtual-node"},
-							},
-							{
-								Key:      "kubernetes.azure.com/cluster",
-								Operator: v1.NodeSelectorOpExists,
-							},
-						},
-					}},
-				},
-			},
-		}
-	}
-
-	// add the user-specified typha preferred affinity if specified.
 	if c.installation.TyphaAffinity != nil && c.installation.TyphaAffinity.NodeAffinity != nil {
-		// check if above code initialized affintiy or not.
+		aff = &v1.Affinity{NodeAffinity: &v1.NodeAffinity{}}
+		// only set if specified to not override the defaults.
+		if c.installation.TyphaAffinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+			aff.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = c.installation.TyphaAffinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+		}
 		// this ensures we still return nil if neither condition is hit.
-		if aff == nil {
-			aff = &v1.Affinity{NodeAffinity: &v1.NodeAffinity{}}
+		if len(c.installation.TyphaAffinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution) > 0 {
+			aff.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution = c.installation.TyphaAffinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution
```

Comment:
The `ensure we still return nil` comment is wrong currently and with the moving of the block above for ProviderAKS `RequiredDuring...` being changed to a default, I think perhaps this code in this block should be adjusted a bit.
I think maybe it should change to
```
// this ensures we return nil if no affinity is specified.
if c.installation.TyphaAffinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil && len(c.installation.TyphaAffinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution) == 0 {
    return nil
}
aff = &v1.Affinity{NodeAffinity: &v1.NodeAffinity{
        RequiredDuringSchedulingIgnoredDuringExecution: c.installation.TyphaAffinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution,
        PreferredDuringSchedulingIgnoredDuringExecution: c.installation.TyphaAffinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution
    }
}
```
Just to be clear it should still include the `if c.installation.TyphaAffinity != nil && c.installation.TyphaAffinity.NodeAffinity != nil {` around the highlighted block.

Reply:
Hey, thanks for comment @tmjd! Addressed as requested.